demo/queued_agent_win.cc

```diff
@@ -0,0 +1,270 @@
+  std::ifstream file(file_path,
+                     std::ios::in | std::ios::binary | std::ios::ate);
+  if (!file.is_open())
+    return false;
+
+  // Get file size.  This example does not handle files larger than 1MB.
+  // Make sure content string can hold the contents of the file.
+  int size = file.tellg();
+  if (size > 1024 * 1024)
+    return false;
+
+  content->resize(size + 1);
+
+  // Read file into string.
+  file.seekg(0, std::ios::beg);
+  file.read(&(*content)[0], size);
+  content->at(size) = 0;
+  return true;
+}
+
+bool ShouldBlockRequest(const std::string& content) {
+  // Determines if the request should be blocked.  For this simple example
+  // the content is blocked if the string "block" is found.  Otherwise the
+  // content is allowed.
+  return content.find("block") != std::string::npos;
+}
+
+// Analyzes one request from Google Chrome and responds back to the browser
+// with either an allow or block verdict.
+void AnalyzeContent(std::unique_ptr<Session> session) {
+  // A session represents one content analysis request and response triggered
+  // by a user action in Google Chrome.  The agent determines whether the
+  // user is allowed to perform the action by examining session->GetRequest().
+  // The verdict, which can be "allow" or "block" is written into
+  // session->GetResponse().
+
+  DumpRequest(session->GetRequest());
+
+  bool block = false;
+  bool success = true;
+
+  if (session->GetRequest().has_content_data() &&
+      session->GetRequest().content_data().has_text_content()) {
+    block = ShouldBlockRequest(
+        session->GetRequest().content_data().text_content());
+  } else if (session->GetRequest().has_content_data() &&
+             session->GetRequest().content_data().has_file_path()) {
+    std::string content;
+    success =
+        ReadContentFromFile(session->GetRequest().content_data().file_path(),
+                            &content);
+    if (success) {
+      block = ShouldBlockRequest(content);
+    }
+  }
+
+  if (!success) {
+    content_analysis::sdk::UpdateResponse(session->GetResponse(), std::string(),
+        ContentAnalysisResponse::Result::FAILURE);
+    std::cout << "  Verdict: failed to reach verdict" << std::endl;
+  } else if (block) {
+    content_analysis::sdk::SetSessionVerdictToBlock(session.get());
+    std::cout << "  Verdict: block" << std::endl;
+  } else {
+    std::cout << "  Verdict: allow" << std::endl;
+  }
+
+  std::cout << std::endl;
+
+  // Send the response back to Google Chrome.
+  if (session->Send() != 0) {
+    std::cout << "[Demo] Error sending response" << std::endl;
+  }
+}
+
+unsigned _stdcall ProcessRequests(void* queue) {
+  RequestQueue* request_queue = reinterpret_cast<RequestQueue*>(queue);
+
+  while (true) {
+    auto session = request_queue->pop();
+    if (!session)
+      break;
+
+    AnalyzeContent(std::move(session));
+  }
+
+  return 0;
+}
+
+int main(int argc, char* argv[]) {
```

Comment:
the pedant in me would make this extern "C"

Reply:
lol main() is a very special function and the compiler and linker [treat it very specially](https://en.cppreference.com/w/cpp/language/main_function).  From that page:
```
a function called "main" cannot be declared with C [language linkage](https://en.cppreference.com/w/cpp/language/language_linkage) in any namespace
```

Reply:
well, i'll be damned! C++-17 says "Its type shall have C++ language linkage and it shall have a declared return type of type int..."